### PR TITLE
feat(assets): support drop-in docker replacements by setting `$CDK_DOCKER`

### LINKED ISSUE #21836: (assets): support drop-in docker replacements
### Describe the feature

I'd like the ability to change the docker command used when building/publishing assets.

### Use Case

I'd like to use [Depot](https://depot.dev/)'s Docker shim as a drop-in replacement for the `docker` command so I can take advantage of their faster docker builds.

### Proposed Solution

We could obtain a user-specified override from an environment variable named `CDK_DOCKER` here:

https://github.com/aws/aws-cdk/blob/7472fa484e45cc5dd05c10e4998c02d28f60da65/packages/cdk-assets/lib/private/docker.ts#L134

Environment variables would be nice because these are easy to add to the build/deploy environment without changing customers' apps. There's precedence for this in core with `CDK_DOCKER`, but that environment variable only applies to bundling and not to docker image asset builds/publishing.

https://github.com/aws/aws-cdk/blob/3e55092fb70e0ec74ee7c4144d6e39a29d8757ae/packages/%40aws-cdk/core/lib/bundling.ts#L523

### Other Information

I've attempted to add this feature in #21838

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

N/A

### Environment details (OS name and version, etc.)

Linux

### PR DESCRIPTION
**Pull request by misterjoshua**
This PR introduces the ability for users to change the command used for building and publishing Docker images to a drop-in replacement like the [Depot](https://depot.dev) Docker shim. To change the docker command, the user provides the alternative command via an environment variable named `CDK_DOCKER`.

There's already something similar in `core`'s asset bundling, but it applies only to bundling and not to building and publishing Docker image assets.

https://github.com/aws/aws-cdk/blob/3e55092fb70e0ec74ee7c4144d6e39a29d8757ae/packages/%40aws-cdk/core/lib/bundling.ts#L523

Fixes #21836 

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

Depends-On: https://github.com/aws/aws-cdk/pull/22194

### CODE REVIEW

**Comment by Hi-Fi** [comment]
How about getting that custom executable to replace `docker` to asset running image? Should custom image be used, some installation added to steps or what? I think this should be mentioned in documentation.

**Comment by misterjoshua** [comment]
> How about getting that custom executable to replace `docker` to asset step's image(s)? Should custom image be used, some installation added to steps or what? I think this should be mentioned in documentation.

@Hi-Fi Thanks for your comments, but I need more clarification. Do you mean to suggest we include documentation on installing third-party tools alongside the AWS CDK and instructions on using those tools with the AWS CDK? i.e., a tutorial on installing and using Depot to build Docker image assets? (`curl https://... | sh && env CDK_DOCKER=depot-shim cdk deploy`)

**Comment by Hi-Fi** [comment]
> > How about getting that custom executable to replace `docker` to asset step's image(s)? Should custom image be used, some installation added to steps or what? I think this should be mentioned in documentation.
> 
> @Hi-Fi Thanks for your comments, but I need more clarification. Do you mean to suggest we include documentation on installing third-party tools alongside the AWS CDK and instructions on using those tools with the AWS CDK? i.e., a tutorial on installing and using Depot to build Docker image assets? (`curl https://... | sh && env CDK_DOCKER=depot-shim cdk deploy`)

As asset stage is a bit special in pipelines, and it's configuration is just one that is [offered](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.pipelines.CodePipeline.html#assetpublishingcodebuilddefaults). So documentation should have recommendation if to use custom build image with executable included or add installation command to some `assetPublishingCodeBuildDefaults`.

Using cdk deploy is different thing, as with that there's no need to worry about stages. And in that I don't think it's needed to tell how to install tool needed.

**Comment by misterjoshua** [comment]
> > > How about getting that custom executable to replace `docker` to asset step's image(s)? Should custom image be used, some installation added to steps or what? I think this should be mentioned in documentation.
> > 
> > 
> > @Hi-Fi Thanks for your comments, but I need more clarification. Do you mean to suggest we include documentation on installing third-party tools alongside the AWS CDK and instructions on using those tools with the AWS CDK? i.e., a tutorial on installing and using Depot to build Docker image assets? (`curl https://... | sh && env CDK_DOCKER=depot-shim cdk deploy`)
> 
> As asset stage is a bit special in pipelines, and it's configuration is just one that is [offered](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.pipelines.CodePipeline.html#assetpublishingcodebuilddefaults). So documentation should have recommendation if to use custom build image with executable included or add installation command to some `assetPublishingCodeBuildDefaults`.
> 
> Using cdk deploy is different thing, as with that there's no need to worry about stages. And in that I don't think it's needed to tell how to install tool needed.

@Hi-Fi Thanks for clarifying! I'll take a crack at adding documentation to the CDK Pipelines README as well.

**Comment by misterjoshua** [comment]
@mergifyio update

File: packages/@aws-cdk/pipelines/README.md [hunk 1]
```diff
@@ -1563,6 +1563,87 @@ We recommend you avoid specifying the `cliVersion` parameter at all. By default
...
+    }),
+    commands: [
+      'npm ci',
+      'npm run build',
+      'npx cdk synth',
+    ],
+  }),
+
+  // Configure CodeBuild to use a drop-in Docker replacement.
+  codeBuildDefaults: {
```
[the remark this answers is not in the document]
Comment (misterjoshua) [reply packages/@aws-cdk/pipelines/README.md:1593]: True that the prop you mention applies only to the publishing steps. But, I'm inclined to think that if a customer is trying to use a drop-in Docker replacement like the Depot Docker shim, they'd first think of using it throughout their CDK app, not just in the asset publishing stage.

For instance, it's not obvious to the caller that `aws_lambda.Code.fromDockerBuild()` builds the Docker image in synth and not during asset publishing - an example with `assetPublishingCodeBuildDefaults` fails to cover this case and could result in unexpected behaviour in this light.

That said, `assetPublishingCodeBuildDefaults` is still available if the customer has a preference.

File: packages/@aws-cdk/pipelines/README.md [hunk 2]
```diff
@@ -1563,6 +1563,87 @@ We recommend you avoid specifying the `cliVersion` parameter at all. By default
...
+
+By default, the AWS CDK will build and publish Docker image assets using the
+`docker` command. However, by specifying the `CDK_DOCKER` environment variable,
+you can override the command that will be used to build and publish your
+assets.
+
+In CDK Pipelines, the drop-in replacement for the `docker` command must be
+included in the CodeBuild environment and configured for your pipeline:
+
+**Adding to the default CodeBuild image**
```
Comment (rix0rrr) [inline packages/@aws-cdk/pipelines/README.md:1576]:
```suggestion
### Adding to the default CodeBuild image
```

File: packages/@aws-cdk/pipelines/README.md [hunk 3]
```diff
@@ -1563,6 +1563,87 @@ We recommend you avoid specifying the `cliVersion` parameter at all. By default
...
+      }
+    }
+  },
+
+  // Turn this on if the application uses bundled file assets
+  dockerEnableForSynth: true,
+});
+```
+
+**Using a custom build image**
```
Comment (rix0rrr) [inline packages/@aws-cdk/pipelines/README.md:1615]: 👍 on including an example for this, but it feels repetitive and doesn't have a paragraph explaining what it is you're doing or why.

Can you add a descriptive paragraph, and cut out nonessentials from the example? For example, declare the source with `declarse const source: CodePipelineSource`, pass a symbolic commands array, that kind of thing. Let's try to leave only the most important parts to highlight them.

File: packages/@aws-cdk/pipelines/README.md [hunk 4]
```diff
@@ -1563,6 +1563,87 @@ We recommend you avoid specifying the `cliVersion` parameter at all. By default
...
+assets.
+
+In CDK Pipelines, the drop-in replacement for the `docker` command must be
+included in the CodeBuild environment and configured for your pipeline:
+
+**Adding to the default CodeBuild image**
+
+```ts
+const pipeline = new pipelines.CodePipeline(this, 'Pipeline', {
+  selfMutation: false,
```
Comment (rix0rrr) [inline packages/@aws-cdk/pipelines/README.md:1580]: Please don't recommend this :).

**Review by rix0rrr — CHANGES_REQUESTED** [review]
I think the `CDK_DOCKER` feature was mostly intended for testing, but I guess I don't mind supporting this. A couple of doc updates and then this is good to go.

(Sidebar: the more I see and learn about Docker, the less I like supporting it. It seems everyone has very specific opinions on how it should be used in their specific use cases, leading to a proliferation of flags and tweaks that are tough to maintain... 😔)

**Comment by misterjoshua** [comment]
@mergifyio update

**Comment by misterjoshua** [comment]
@mergifyio update

**Comment by rix0rrr** [comment]
Needs https://github.com/aws/aws-cdk/pull/22194 to be merged first

**Comment by rix0rrr** [comment]
@Mergifyio update

**Pull request merged**